boards/arm/adafruit_feather_stm32f405/board.cmake

```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=STM32F405RG" "--speed=4000")
```

Comment:
I just found this pull request after seeing the blog post about this board on Slack (https://blog.adafruit.com/2020/02/26/new-guide-blinking-an-led-with-the-zephyr-rtos-adafruitlearningsystem-zephyr-adafruit-zephyriot-adafruit/).

The dfu-util runner worked fine when I last tested it:

https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/west_commands/runners/dfu.py

It's still in use in a mainline board:

https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/96b_carbon/board.cmake

And you can have multiple runners in one board.cmake:

https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf52_pca10040/board.cmake

(The first `include` sets the default, but users can override with e.g. `west flash -r SOME_OTHER_NON_DEFAULT_RUNNNER`).

It might be nice to see this PR updated to add dfu-util support for this board. That would let you update the blog post so your users can just use west instead of having to call out to dfu-util manually.

Reply:
Thanks for the link and examples. I'll take a look at testing this today.

Reply:
@mbolivar I'm running into some roadblocks. I'm admittedly not an expert in DFU Util, but how are you linking the image, exactly? It is not defined in the boards/common file. Additionally, why is pid a required parameter? And is there a reason that none of the options can use their short form?

Reply:
I see now that this is a custom interface to a python script. I see no way of specifying the address in your script - this is a requirement for STM32 devices, which start at 0x8000000. Is this implemented in a way I'm not recognizing?

Reply:
Well, it's working anyway.

Reply:
> I see no way of specifying the address in your script - this is a requirement for STM32 devices, which start at 0x8000000. Is this implemented in a way I'm not recognizing?

The address is taken care of by the `--dfuse` option you set in board.cmake.

You can run `west --verbose flash` to see the dfu-util command itself.

Reply:
BTW, the documentation is here: https://docs.zephyrproject.org/latest/guides/west/build-flash-debug.html


If anything is missing or could be improved, please let me know! Thanks.

Reply:
@mbolivar is that dfuse option completely automatic? Is it possible to change it to a custom value? Sorry for pestering, I don't see `dfuse` mentioned in the documentation, and I'm not fluent in python. While the flash process is working now, I'd appreciate knowing the option for bootloader settings, etc. 

My only requests for this might be that the interface be brought closer to the [man page documentation](http://dfu-util.sourceforge.net/dfu-util.1.html), supporting shorthand options, allowing use of `--dfuse-address`, etc. It'd also be nifty if your runner documentation was linked in the [porting](https://docs.zephyrproject.org/latest/guides/porting/board_porting.html) section. I think that whole section could use a bit of an overhaul to offer more guidance for adding boards, but that's just my first impression.

Reply:
In any case thanks for your help - dfu-util is working now, and it looks like it's passed everything but this one unrelated timeout issue. I'll try and update the guide this week.

Reply:
> Sorry for pestering, I don't see `dfuse` mentioned in the documentation, and I'm not fluent in python. 

You're not pestering :). Some of these details aren't exactly as I would wish them to be.

Regarding help, you can run `west help flash` for common flash options, and `west flash --context --runner dfu-util` (or `west flash -H -r dfu-util` for short) for dfu-util specific options.

If `--context` has a build directory to work with, it will emit extra output specific to that build directory.

In this case:

```
  --dfuse               set if target is a DfuSe device; implies --dt-flash.
  --dfuse-modifiers DFUSE_MODIFIERS
                        colon-separated list of DfuSe modifiers (default is "leave", which starts
                        execution immediately); --dfuse must also be given for this option to take
                        effect.

```

There is no short option for these, unlike e.g. `--runner` which has `-r`, which you can see in the `west help flash` output:

```
  -r RUNNER, --runner RUNNER
                        overrides the default flash runner
```

I would be against adding one, because these are set once in board.cmake instead of typed out every time, and IMO it's best to use long options when they're being saved in a source file for readability.

> @mbolivar is that dfuse option completely automatic? Is it possible to change it to a custom value? 

--dfuse doesn't take any arguments (see help above).

The "dfuse" name is dfu-util-ese for DFU + STM32 Extensions (http://dfu-util.sourceforge.net/dfuse.html).

There should be no need to change it under ordinary circumstances, because when --dfuse is given, the flash address is determined by CONFIG_FLASH_BASE_ADDRESS (and CONFIG_FLASH_LOAD_OFFSET if that is set).  The command line help text for --dfuse could be clearer about that. I will fix it.

For example, with your board and this PR, I get this in the `west -v flash` output:

```
runners.dfu-util: dfu-util -d,0483:df11 -s 0x8000000:leave -a 0 -D /home/mbolivar/zp/zephyr/build/zephyr/zephyr.bin
```

The 0x8000000 comes from Kconfig, so there should be no reason to change it, since it's being flashed to the address you asked Zephyr to link your application into. If that doesn't cover a use case you need, please let me know.

Reply:
> My only requests for this might be that the interface be brought closer to the [man page documentation](http://dfu-util.sourceforge.net/dfu-util.1.html), supporting shorthand options, allowing use of `--dfuse-address`, etc.

The way to get help out of west is currently just via --help and related options. Adding manpages is something I've thought about, but would be difficult to make work on Windows (even things like `git help` open web browsers on Windows).

You can add extra modifiers (that go onto the `-s` dfu-util option) via the `--dfuse-modifiers` mentioned in the help output I pasted above.

If you have a use case for overriding the flash address that you can't get done just by using a Kconfig overlay that picks the CONFIG_FLASH_BASE_ADDRESS you want, please open an issue and I'm happy to discuss.

> It'd also be nifty if your runner documentation was linked in the [porting](https://docs.zephyrproject.org/latest/guides/porting/board_porting.html) section. I think that whole section could use a bit of an overhaul to offer more guidance for adding boards, but that's just my first impression.

I agree the porting guide has lots of issues, unfortunately :(. Let's fix this one while we're here, indeed: https://github.com/zephyrproject-rtos/zephyr/pull/23205

Reply:
I appreciate that you've thought all this out. From my outside perspective, the reason I was interested in a compatible interface to true `dfu-util` was because my first instinct on porting the board is to simply replicate the command I normally use, `dfu-util -a 0 --s 0x08000000 -D zephyr.bin` in this case. I understand wanting to keep things readable - do you think it'd be appropriate to add a line to the docs or command return that informs would-be porters that a) flash address and binary linking are done automatically and b) long-form options are required as parameters?

Reply:
> do you think it'd be appropriate to add a line to the docs or command return that informs would-be porters that  a) flash address and binary linking are done automatically and

The "flash address" part I think is addressed in https://github.com/zephyrproject-rtos/zephyr/pull/23205 -- agreed? If not, let's fix it.

I don't know what you mean by "add a line to the docs or command return that informs would-be porters that [...] binary linking [is] done automatically", though.

That sounds like a question about the build system documentation. The way the build system links zephyr.elf is very architecture-specific and there are numerous Kconfig options which affect its behavior, but they should all have `help` text and https://docs.zephyrproject.org/latest/application/index.html#cmake-details documents the use of Kconfig in general.

Is there something specific you'd like to see here?

> b) long-form options are required as parameters?

I don't know what you mean here either, sorry. Some west command line options have short forms, and some don't. You can always tell by running `west help flash`, or in the case of the runners, `west flash --context -r <runner>`.

In terms of discoverability:

- the `west flash --context` option is printed in the `west help flash` output
- the suggestion to run `west help <command>` (so e.g. `west help flash`) is printed by all of these: `west --help`, `west -h`, `west`, `west help` (i.e. west supports all of the commonly used "give me some help" idioms)

The top page of the west docs (https://docs.zephyrproject.org/latest/guides/west/index.html) says:

> You can run west --help (or west -h for short) to get top-level help for available west commands, and west <command> -h for detailed help on each command.

So other than copy/pasting the existing command line options into the documentation, which would in practice go stale instantly because I doubt I could get contributors to maintain the extra copy, I'm genuinely not sure what I can do here.

But I have an insider perspective so I'm probably missing something. If you can suggest a concrete change to the docs I would be grateful.

Thanks.

Reply:
@mbolivar I should have been more clear - I don't mean any changes to the options or operation of west! I simply mean that when I started, I was confused by the differences between the dfu-util python runner and actual dfu-util, and I didn't fully understand what was up I found and read the dfu-util runner script. EG, you can't type shorthand `board_runner_args(dfu-util "--a=0" "-s=0x8000000")`, the PID:VID pair is required where it isn't normally, and the binary is selected automatically by the script. Even in the helpful changes you made to the porting guide, it isn't specified that a runner is actually a python script intermediary that has a different interface than the bash tool it is named after. 

I feel it might be helpful if these specifics were documented somewhere, but maybe it's more obvious than I'm giving it credit for and it just flew over my head in this instance.

Reply:
Specific addition might be 

> To add west flash and west debug support, you need to add a board.cmake file to your board directory which configures a “runner” python script for your board. (There’s nothing special you need to do to get west build support for your board.) "Runner" scripts select the zephyr binary automatically, and may may have a different interface from the utilities they wrap around for readability and to match Zephyr requirements - run `west flash --context` to see available runners and `west flash --context -r RUNNER` to view the specific requirements for each runner.

Reply:
> Specific addition might be

Awesome, thank you so much for that. I've tried to fold that in here: https://github.com/zephyrproject-rtos/zephyr/pull/23266

Comments welcome!